### DISCUSSION TITLE
[Feature Request] Add `SortDirection` parameter to custom `SortingFn`

### DISCUSSION THREAD

**Discussion by rpiaggio** [TanStack/table#4920] [Ideas]
We have cases where we want certain values to always be sorted at the end despite of sort direction.

This can be achieved by passing the column sort direction to a custom sorting function, where we can treat this case.

Eg, have `SortingFn`'s signature be:

``` js
export type SortingFn<TData extends AnyData> = {
  (rowA: Row<TData>, rowB: Row<TData>, columnId: string, direction: SortDirection): number
}
```

It seems this could help other users too (https://github.com/TanStack/table/discussions/3300).

Thank you!

**Comment by jodoox** [comment]
@tannerlinsley Huge fan of your libraries, thanks for all these contributions.

Having undefined values at the bottom of the list regardless of the direction is a reasonable need, yet I believe it is not possible to achieve it easily with the custom sorting functions as of now. Exposing the sort direction seems like a simple yet useful addition that would solve this issue, would you be open to a PR for this?

**Reply by artchen-db** [reply]
+1 I too think sorting "null values" always at bottom is a reasonable feature to support, and passing sort direction to the sorting function is essential to implementing this feature.

**Reply by jodoox** [reply]
I solved it for the moment with this helper `sortingFn: categoricalSortingFn(table, ["high", "medium", "low"])`
Got no answer [on the discord](https://discord.com/channels/719702312431386674/1173667691060543628) as well
```
export const categoricalSortingFn = <RowType extends Record<string, unknown>>(
  table: Table<RowType> | null,
  values: string[],
): SortingFn<RowType> => {
  return (rowA: Row<RowType>, rowB: Row<RowType>, columnId: string) => {
    const desc = table?.getState().sorting.find((x) => x.id === columnId)?.desc;
    const sortedArray = desc ? [...values, undefined] : [undefined, ...values];
    const valA = rowA.getValue<string | undefined>(columnId);
    const valB = rowB.getValue<string | undefined>(columnId);
    const iA = sortedArray.indexOf(valA);
    const iB = sortedArray.indexOf(valB);
    const order = iA < iB ? 1 : -1;
    return order;
  };
};

```

**Reply by artchen-db** [reply]
Thanks! This approach of passing a ref to the table model works great in our use case too.

**Reply by rpiaggio** [reply]
A couple of month ago https://github.com/TanStack/table/pull/5486 was merged, which allows `undefined` values to always be pulled first or last when sorting, regardless of direction.

**Reply by marceloverdijk** [reply]
@rpiaggio but this `sortUndefined`: `last` option does not support `null` values unfortunately.

**Reply by marceloverdijk** [reply]
I'm doing something like this now:

```
export const nullAwareNumber: SortingFn<any> = (a, b, columnId) => {
  const col = a.getAllCells().find(c => c.column.id === columnId)?.column
  const nullsLast = col?.columnDef.meta?.nullsLast ?? false

  const x = a.getValue<number | null>(columnId)
  const y = b.getValue<number | null>(columnId)

  if (x == null && y == null) return 0
  if (nullsLast) {
    if (x == null) return 1
    if (y == null) return -1
  } else {
    if (x == null) return -1
    if (y == null) return 1
  }
  return (x ?? 0) - (y ?? 0)
}

```

but I have to "find".the column for each row this way, not sure if this will give a noticeable performance penalty with large datasets...

**Reply by marceloverdijk** [reply]
and now I'm doing this everywhere:

```
columnHelper.accessor(
        row => row.someColumnWithNullValues ?? undefined,
        {
          id: 'someColumn',
          header: 'Header',
          sortDescFirst: true,
          sortUndefined: 'last',
        },
      ),
```

**Comment by artchen-db** [comment]
I was reading some old issues like https://github.com/TanStack/table/discussions/2371, it seems the direction was part of the function signature. It would be great if we could bring it back.

**Comment by sullyslegacy** [comment]
The sortEntry has id and desc however only id is being passed to the sortingFn. It would be very helpful if desc was passed in as well.

However a workaround would be doing something like
`
sortingFn: (obj1: Row<MyType>, obj2: Row<MyType>, prop: string) =>
                sortUndefinedAtBottom(
                    customValueFunction(obj1.original),
                    customValueFunction(obj2.original),
                    isColumnSortedDesc(obj1, prop)
                ),

const isColumnSortedDesc = <T,>(column: Row<T>, prop: string) => {
    const sort = column._getAllCellsByColumnId()[prop].column.getIsSorted();
    return sort === ('desc' as SortDirection);
};

const sortUndefinedAtBottom = (value1: number | undefined, value2: number | undefined, isSortedDesc: boolean) => {
    if (value1 === value2) {
        return 0;
    }

    if (value1 === undefined) {
        return isSortedDesc ? -1 : 1;
    } else if (value2 === undefined) {
        return isSortedDesc ? 1 : -1;
    }

    return value1 - value2;
};
`

**Comment by inomn** [comment]
Hi! In case you want to sort columns with range values, for example  [241, 567], and for ascending I would like to compare rows by 1st value (241), but for descending by last (567).

It is not possible to write a custom sorting function to achieve that without having `sortDirection`.
I would appreciate adding `sortDirection` to `sortingFn` signature.

**Reply by shahbazshueb** [reply]
Hi. I need a specific value to be sorted at the end so really need this feature.
+1 On this. Would really appreciate if sortingFn is added to sortDirection.

**Comment by ippeiukai** [comment]
Just wanted to note related issues and discussions on sortingFn vs direction topic:
- issues
  - https://github.com/TanStack/table/pull/5486
  - https://github.com/TanStack/table/issues/5653
  - https://github.com/TanStack/table/issues/6061
- discussions
  - https://github.com/TanStack/table/discussions/2371
  - https://github.com/TanStack/table/discussions/3300
  - https://github.com/TanStack/table/discussions/5453

**Reply by ippeiukai** [reply]
Also noting that react-table v7 did have `desc` as the 4th param of `SortingFn`:
https://react-table-v7-docs.netlify.app/docs/api/usesortby#column-options
> `sortType: String | Function(rowA: <Row>, rowB: <Row>, columnId: String, desc: Bool)`
